src/ImportDefinitionsBundle/Provider/ExcelProvider.php

```diff
@@ -0,0 +1,199 @@
+<?php
+/**
+ * Import Definitions.
+ *
+ * LICENSE
+ *
+ * This source file is subject to the GNU General Public License version 3 (GPLv3)
+ * For the full copyright and license information, please view the LICENSE.md and gpl-3.0.txt
+ * files that are distributed with this source code.
+ *
+ * @copyright  Copyright (c) 2016-2018 w-vision AG (https://www.w-vision.ch)
+ * @license    https://github.com/w-vision/ImportDefinitions/blob/master/gpl-3.0.txt GNU General Public License version 3 (GPLv3)
+ */
+
+namespace ImportDefinitionsBundle\Provider;
+
+use Box\Spout\Common\Type;
+use Box\Spout\Reader\ReaderFactory;
+use Box\Spout\Reader\ReaderInterface;
+use Box\Spout\Writer\WriterFactory;
+use Box\Spout\Writer\WriterInterface;
+use Box\Spout\Writer\XLSX\Writer;
+use ImportDefinitionsBundle\Model\ExportDefinitionInterface;
+use ImportDefinitionsBundle\Model\ImportMapping\FromColumn;
+use ImportDefinitionsBundle\ProcessManager\ArtifactGenerationProviderInterface;
+use ImportDefinitionsBundle\ProcessManager\ArtifactProviderTrait;
+use Pimcore\Model\Asset;
+
+class ExcelProvider implements ProviderInterface, ExportProviderInterface, ArtifactGenerationProviderInterface
+{
+    use ArtifactProviderTrait;
+
+    /** @var string */
+    private $exportPath;
+
+    /** @var WriterInterface */
+    private $writer;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function testData($configuration)
+    {
+        return true;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getColumns($configuration)
+    {
+        if ($configuration['exampleFile']) {
+            $exampleFile = Asset::getById($configuration['exampleFile']);
+            if (null !== $exampleFile) {
+                $reader = $this->createReader(PIMCORE_ASSET_DIRECTORY . $exampleFile->getFullPath());
+
+                $sheetIterator = $reader->getSheetIterator();
+                $sheetIterator->rewind();
+                $rowIterator = $sheetIterator->current()->getRowIterator();
+                if (null !== $rowIterator) {
+                    $rowIterator->rewind();
+
+                    return $this->buildColumns($rowIterator->current());
+                }
+            }
+        } elseif ($configuration['excelHeaders']) {
+            return $this->buildColumns(str_getcsv($configuration['excelHeaders']));
+        }
+
+        return [];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getData($configuration, $definition, $params, $filter = null)
+    {
+        $file = sprintf('%s/%s', PIMCORE_PROJECT_ROOT, $params['file']);
+
+        $reader = $this->createReader($file);
+        $sheetIterator = $reader->getSheetIterator();
+        $sheetIterator->rewind();
+        $rowIterator = $sheetIterator->current()->getRowIterator();
+        
+        return (function (\Iterator $iterator): \Generator {
```

Comment:
Returning a generator will stream the data from the Excel file instead of building an in-memory representation.

Reply:
Thanks to @ciaranmcnulty